libs/Microsoft.MixedReality.WebRTC.Unity/Assets/Microsoft.MixedReality.WebRTC.Unity.Tests.Runtime/VideoSourceTests.cs

```diff
@@ -0,0 +1,1110 @@
+            var pc1 = pc1_go.AddComponent<PeerConnection>();
+            pc1.AutoInitializeOnStart = false;
+            var pc2_go = new GameObject("pc2");
+            pc2_go.SetActive(false); // prevent auto-activation of components
+            var pc2 = pc2_go.AddComponent<PeerConnection>();
+            pc2.AutoInitializeOnStart = false;
+
+            // Create the signaler
+            var sig_go = new GameObject("signaler");
+            var sig = sig_go.AddComponent<HardcodedSignaler>();
+            sig.Peer1 = pc1;
+            sig.Peer2 = pc2;
+
+            // Create the video sources on peer #1
+            var sender1 = pc1_go.AddComponent<FakeVideoSource>();
+            sender1.AutoStartOnEnabled = true;
+            sender1.TrackName = "track_name";
+            var receiver1 = pc1_go.AddComponent<VideoReceiver>();
+            MediaLine tr1 = pc1.AddTransceiver(MediaKind.Video);
+            tr1.Sender = sender1;
+            tr1.Receiver = receiver1;
+
+            // Create the video sources on peer #2
+            var sender2 = pc2_go.AddComponent<FakeVideoSource>();
+            sender2.AutoStartOnEnabled = true;
+            sender2.TrackName = "track_name";
+            var receiver2 = pc2_go.AddComponent<VideoReceiver>();
+            MediaLine tr2 = pc2.AddTransceiver(MediaKind.Video);
+            tr2.Sender = sender2;
+            tr2.Receiver = receiver2;
+
+            // Activate
+            pc1_go.SetActive(true);
+            pc2_go.SetActive(true);
+
+            // Initialize
+            var initializedEvent1 = new ManualResetEventSlim(initialState: false);
+            pc1.OnInitialized.AddListener(() => initializedEvent1.Set());
+            Assert.IsNull(pc1.Peer);
+            pc1.InitializeAsync().Wait(millisecondsTimeout: 50000);
+            var initializedEvent2 = new ManualResetEventSlim(initialState: false);
+            pc2.OnInitialized.AddListener(() => initializedEvent2.Set());
+            Assert.IsNull(pc2.Peer);
+            pc2.InitializeAsync().Wait(millisecondsTimeout: 50000);
+
+            // Wait a frame so that the Unity event OnInitialized can propagate
+            yield return null;
+
+            // Check the event was raised
+            Assert.IsTrue(initializedEvent1.Wait(millisecondsTimeout: 50000));
+            Assert.IsNotNull(pc1.Peer);
+            Assert.IsTrue(initializedEvent2.Wait(millisecondsTimeout: 50000));
+            Assert.IsNotNull(pc2.Peer);
+
+            // Confirm the senders are ready
+            Assert.NotNull(sender1.Track);
+            Assert.NotNull(sender2.Track);
+
+            // Connect
+            Assert.IsTrue(sig.Connect(millisecondsTimeout: 60000));
+
+            // Wait a frame so that the Unity events for streams started can propagate
+            yield return null;
+
+            // Check pairing
+            Assert.NotNull(receiver1.Track);
+            Assert.NotNull(receiver2.Track);
+        }
```

Comment:
Making these a parameterized function wouldn't be bad - there are only a few lines that are different between each.

Reply:
Also moving some of the blocks in a function/method would help since they seem to be shared by the other tests below.